domain/password/state/state.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+
+	"github.com/canonical/sqlair"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/unit"
+	"github.com/juju/juju/domain"
+	applicationerrors "github.com/juju/juju/domain/application/errors"
+	"github.com/juju/juju/domain/password"
+	"github.com/juju/juju/internal/errors"
+)
+
+// State defines the access mechanism for interacting with passwords in the
+// context of the model database.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState constructs a new state for interacting with the underlying passwords
+// of a model.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// SetUnitPasswordHash sets the password hash for the given unit.
+func (s *State) SetUnitPasswordHash(ctx context.Context, unitUUID unit.UUID, passwordHash password.PasswordHash) error {
+	db, err := s.DB()
+	if err != nil {
+		return err
+	}
+
+	args := unitPasswordHash{
+		UUID:         unitUUID,
+		PasswordHash: passwordHash,
+	}
+
+	query := `
+UPDATE unit
+SET password_hash = $unitPasswordHash.password_hash
+WHERE uuid = $unitPasswordHash.uuid;
+`
+	stmt, err := s.Prepare(query, args)
+	if err != nil {
+		return errors.Errorf("preparing statement to set password hash: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var outcome sqlair.Outcome
+		if err := tx.Query(ctx, stmt, args).Get(&outcome); err != nil {
+			return errors.Errorf("setting password hash: %w", err)
+		}
+		result := outcome.Result()
+		if err != nil {
+			return errors.Errorf("getting result of setting password hash: %w", err)
+		}
+		if affected, err := result.RowsAffected(); err != nil {
+			return errors.Errorf("getting number of affected rows: %w", err)
+		} else if affected == 0 {
+			return applicationerrors.UnitNotFound
```

Comment:
Is this correct? I'm thinking out loud, but we have the password right up when we create the unit?